mit_d3m/__init__.py

```diff
@@ -11,63 +11,98 @@
 import tarfile
 
 import boto3
+import botocore
+import botocore.config
+from funcy import memoize
 
 from mit_d3m.dataset import D3MDS
 from mit_d3m.loaders import get_loader
 from mit_d3m.metrics import METRICS_DICT
+from mit_d3m.utils import contains_files
+
+__all__ = (
+    'DATA_PATH',
+    'BUCKET',
+    'load_d3mds',
+    'load_dataset',
+)
+
 
 DATA_PATH = 'data'
 BUCKET = 'd3m-data-dai'
 
 
-def download_dataset(bucket, dataset, root_dir):
-    client = boto3.client('s3')
+@memoize
+def get_client():
+    config = botocore.config.Config(signature_version=botocore.UNSIGNED)
+    client = boto3.client('s3', config=config)
+    return client
+
+
+def get_dataset_tarfile_path(datapath, dataset):
+    return os.path.join(datapath, '{dataset}.tar.gz'.format(dataset=dataset))
+
 
-    print("Downloading dataset {}".format(dataset))
+def get_dataset_dir(datapath, dataset):
+    return os.path.join(datapath, dataset)
 
-    key = 'datasets/' + dataset + '.tar.gz'
-    filename = root_dir + '.tar.gz'
 
-    print("Getting file {} from S3 bucket {}".format(key, bucket))
+def get_dataset_s3_key(dataset):
+    return 'datasets/{dataset}.tar.gz'.format(dataset=dataset)
+
+
+def download_dataset(bucket, key, filename):
+    print("Downloading dataset from s3://{bucket}".format(bucket=bucket))
+    client = get_client()
     client.download_file(Bucket=bucket, Key=key, Filename=filename)
 
-    shutil.rmtree(root_dir, ignore_errors=True)
 
-    print("Extracting {}".format(filename))
-    with tarfile.open(filename, 'r:gz') as tf:
-        tf.extractall(os.path.dirname(root_dir))
+def extract_dataset(src, dst):
+    print("Extracting {}".format(src))
+    shutil.rmtree(dst, ignore_errors=True)
+    with open(src, 'r') as f:
+        tf = tarfile.TarFile(mode='r:gz', fileobj=f)
+        tf.extractall(dst)
 
 
 def load_d3mds(dataset, root=DATA_PATH, force_download=False):
+    if not os.path.exists(root):
+        os.makedirs(root)
+
     if dataset.endswith('_dataset_TRAIN'):
-        dataset = dataset[:-14]
+        dataset = dataset[:len('_dataset_TRAIN')]
 
-    root_dir = os.path.join(root, dataset)
+    dataset_dir = get_dataset_dir(root, dataset)
+    dataset_tarfile = get_dataset_tarfile_path(dataset_dir, dataset)
 
-    if root == DATA_PATH and (force_download or not os.path.exists(root_dir)):
-        if not os.path.exists(root_dir):
-            os.makedirs(root_dir)
+    if force_download or not os.path.exists(dataset_tarfile):
+        download_dataset(BUCKET, dataset, dataset_tarfile)
 
-        download_dataset(BUCKET, dataset, root_dir)
+    if force_download or not os.path.exists(dataset_dir) or not contains_files(dataset_dir):
```

Comment:
I do not see the point of this `contains_files` here. In which case one would end up having the dataset folder there but without any regular files? And if this happened, why not expect the user to indicate `force_download` to overwrite it?

Reply:
as I wrote in my longer comment, it indicates an error condition in which there was an error extracting the files. it indicates that we need to retry extracting the tarfile. but force_download would require to re-download the tarfile which may not be necessary. there may be an alternate way to elegantly handle this error condition.